dipy/reconst/rumba.py

```diff
@@ -0,0 +1,815 @@
+        if lambda_csf is not None and lambda_csf < 0:
+            raise ValueError(
+                f"lambda_csf must be None or > 0, received {lambda_csf}")
+
+        self.lambda1 = lambda1
+        self.lambda2 = lambda2
+        self.lambda_csf = lambda_csf
+
+        ## Initializing remaining parameters ##
+        if R < 1 or n_iter < 1 or n_coils < 1:
+            raise ValueError(f"R, n_iter, and n_coils must be >= 1, but R={R}, n_iter={n_iter}" +
+                             f", and n_coils={n_coils} ")
+        if recon_type not in ['smf', 'sos']:
+            raise ValueError(
+                f"Invalid recon_type. Should be 'smf' or 'sos', received {recon_type}")
+
+        self.R = R
+        self.n_iter = n_iter
+        self.recon_type = recon_type
+        self.n_coils = n_coils
+
+    @multi_voxel_fit
+    def fit(self, data):
+
+        ## Signal repair, normalization ##
+
+        # Normalize data to mean b0 image
+        data = normalize_data(data, self.where_b0s, min_signal=EPS)
+        # Rearrange data to match corrected gradient table
+        data = np.concatenate(([1], data[self.where_dwi]))
+        data[data > 1] = 1  # Clip values between 0 and 1
+
+        return RumbaFit(self, data)
+
+
+class RumbaFit(OdfFit):
+
+    def __init__(self, model, data):
+        '''
+        Computes fODF for a single voxel
+
+        Parameters
+        ----------
+        model : RumbaSD
+            RUMBA-SD model.
+        data : 1d ndarray
+            signal values for a single voxel.
+
+        '''
+        OdfFit.__init__(self, model, data)
+        self.kernel = None
+        self._f_csf = None
+
+        self._gfa = None
+        self.npeaks = 5
+        self._peak_values = None
+        self._peak_indices = None
+        self._qa = None
+
+    def odf(self, sphere):
+        '''
+        Computes fODF at discrete vertices on sphere.
+
+        Parameters
+        ----------
+        sphere : Sphere
+            sphere on which to construct fODF.
+
+        Returns
+        -------
+        fodf_wm : 1d ndarray
+            fODF computed at each vertex on `sphere`.
+
+        '''
+
+        ## Creation of kernel ##
+
+        # Check if previously constructed
+        self.kernel = self.model.cache_get('kernel', key=sphere)
+
+        if self.kernel is None:
+            self.kernel = generate_kernel(self.model.gtab, sphere, self.model.lambda1,
+                                          self.model.lambda2, self.model.lambda_csf)
+            self.model.cache_set('kernel', sphere, self.kernel)
+
+        ## Fitting ##
+        fodf_wm, f_csf = rumba_deconv(self.data, self.kernel,
+                                      self.model.n_iter, self.model.recon_type,
+                                      self.model.n_coils)
+
+        self._f_csf = f_csf
+
+        return fodf_wm
+
+    @property
+    def f_csf(self):
+        if self._f_csf is None:
+            raise RuntimeError(
```

Comment:
I think here it would be ideal to move the "fitting" code from the function `.odf(.)` to a class function `._fit(.)`. Then, if `.f_csf(.)` or `.odf(.)` functions are called, you return the corresponding value, and call `_fit(.)` if the value is `None`.